### PR TITLE
split browse view to distinct routes

### PR DESCRIPTION
**Pull request by pgandhizaizi**
## Changes in this PR

1. separated route for each browse view
2. separated test cases and e2e tests for each browse view
3. fixed route locations of each browse view in every possible places in code base
4.  created parameterised test cases for browse views

## JIRA ticket

AYR-620

### After

- [ ] Requires env variable(s) to be updated

### CODE REVIEW

File: app/main/routes.py [hunk 1]
```diff
@@ -92,106 +96,232 @@ def accessibility():
...
+            num_records_found=num_records_found,
+            dict_args={
+                k: v for k, v in request.args.items() if k not in "page"
+            },
+        )
+
+
+@bp.route("/browse/transferring_body/<uuid:_id>", methods=["GET"])
+@access_token_sign_in_required
+def browse_transferring_body(_id: uuid.UUID):
```
Comment (anthonyhashemi) [inline app/main/routes.py:158]: why `_id` and not just `id`?

File: app/main/routes.py [hunk 2]
```diff
@@ -92,106 +96,232 @@ def accessibility():
...
+    """
+    Render the browse transferring body view page.
+
+    This function retrieves search results for a specific
+    record(s) based on the 'transferring_body_id' provided in the query parameters, and renders
```
Comment (anthonyhashemi) [inline app/main/routes.py]:
```suggestion
    record(s) based on the transferring_body 'id' provided
```

as its not from the query params anymore

File: app/main/routes.py [hunk 3]
```diff
@@ -92,106 +96,232 @@ def accessibility():
...
+    # set default sort for consignment view
+    if len(sorting_orders) == 0:
+        sorting_orders["closure_type"] = "asc"
+
+    query = build_browse_consignment_query(
+        consignment_id=_id,
+        filters=filters,
+        sorting_orders=sorting_orders,
+    )
+    print(query.subquery())
```
Comment (anthonyhashemi) [inline app/main/routes.py]: lets remove this print statement

File: app/main/routes.py [hunk 4]
```diff
@@ -92,106 +96,232 @@ def accessibility():
...
+        2: {"series_id": series.SeriesId},
+        3: {"series": series.Name},
+        4: {"consignment_reference": consignment.ConsignmentReference},
+    }
+
+    filters = build_browse_consignment_filters(request.args)
+    sorting_orders = build_sorting_orders(request.args)
+
+    # set default sort for consignment view
+    if len(sorting_orders) == 0:
```
Comment (anthonyhashemi) [inline app/main/routes.py:312]: is there something unique about consingments sorting that means that we need to set the default sort explicitly out here as opposed to inside build_sorting_orders like the other route?

File: app/templates/main/browse-series.html [hunk 5]
```diff
@@ -124,25 +131,26 @@ <h3 class="govuk-heading-s govuk-heading-s--browse-all-filter">Date consignment
...
-                                <a href="{{ url_for('main.browse', consignment_id=record['consignment_id']) }}">{{ record["consignment_reference"] }}</a>
+                                <a href="{{ url_for('main.browse_consignment', _id=record['consignment_id']) }}">{{ record["consignment_reference"] }}</a>
                             </div>
                         </td>
                     </tr>
                 {% endfor %}
             </tbody>
         </table>
+        <!-- PAGINATION -->
+        <div class="govuk-grid-column-two-thirds browse-grid--two-thirds">
```
Comment (anthonyhashemi) [inline app/templates/main/browse-series.html:149]: so were we previously completely missing pagination for all these views?

File: app/templates/main/pagination.html [hunk 6]
```diff
@@ -5,7 +5,7 @@
         {% if results.has_prev == True %}
             <div class="govuk-pagination__prev">
                 <a class="govuk-link govuk-pagination__link"
-                   href="{{ url_for(view_name, page=current_page-1, **filters) }}"
+                   href="{% if id != None %} {{ url_for(view_name, _id=id, page=current_page-1, **dict_args) }} {% else %} {{ url_for(view_name, page=current_page-1, **dict_args) }} {% endif %}"
```
Comment (anthonyhashemi) [inline app/templates/main/pagination.html]: could we rename dict_args something a bit more meaningful as its quite vague at the moment?

File: app/main/routes.py (hunk 1 above)
Comment (pgandhizaizi) [reply app/main/routes.py:158]: id is a reserved keyword in Python and is conflicting when used

File: app/main/routes.py (hunk 4 above)
Comment (pgandhizaizi) [reply app/main/routes.py:312]: as we are using generic sort from filter_sort_builder, if we want a default sorting on specific browse view, we have to define in the route of that view

File: app/templates/main/browse-series.html (hunk 5 above)
Comment (pgandhizaizi) [reply app/templates/main/browse-series.html:149]: pagination was on browse.html for common use, it has been moved to individual browse view to pass in _id parameter based on that view

File: app/templates/main/pagination.html (hunk 6 above)
Comment (pgandhizaizi) [reply app/templates/main/pagination.html]: please suggest the name and I will replace it

File: app/main/routes.py (hunk 2 above)
Comment (pgandhizaizi) [reply app/main/routes.py]: updated comment for transferring body, series and consignment routes as requested

File: app/main/routes.py (hunk 3 above)
Comment (pgandhizaizi) [reply app/main/routes.py]: removed print statement

File: app/templates/main/pagination.html (hunk 6 above)
Comment (pgandhizaizi) [reply app/templates/main/pagination.html]: renamed dict_args to query_string_parameters at all places where dict_args used

File: app/main/routes.py (hunk 1 above)
Comment (anthonyhashemi) [reply app/main/routes.py:158]: Are you sure, I didn't think it was:
Screenshot 2024-02-21 at 14 16 22

File: app/main/routes.py (hunk 2 above)
Comment (anthonyhashemi) [reply app/main/routes.py]: cheers

File: app/main/routes.py (hunk 3 above)
Comment (anthonyhashemi) [reply app/main/routes.py]: thanks

File: app/main/routes.py (hunk 4 above)
Comment (anthonyhashemi) [reply app/main/routes.py:312]: Okay, I noticed the other routes do not set a default sort order. Does that mean we don't care the default order for those other pages? I feel like we definitely need the default sort order of results to match the default sort option in the sort dropdown. Worth double checking this is correct and if anything needs fixing we can create a bug ticket for it

File: app/templates/main/pagination.html (hunk 6 above)
Comment (anthonyhashemi) [reply app/templates/main/pagination.html]: query_string_parameters is much clearer, thank you

File: app/tests/test_browse_consignment.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+                [
+                    [
+                        "'12/04/2023', 'fourth_file.xls', 'Closed', '12/04/2023', "
+                        "'10/03/2023', 'third_file.docx', 'Closed', '10/03/2023'"
+                    ],
+                ],
+            ),
+        ],
+    )
+    def test_browse_consignment_full_test(
```
Comment (anthonyhashemi) [inline app/tests/test_browse_consignment.py:310]: paramterization is helping greatly, makes these tests much easier to follow! Like we discussed, if we get a chance after delivering MVP we may revist to move the testing of edge case of filters and sort combinations to a lower level and leave fewer higher level route tests but already the paramterization is a big help.

**Review by anthonyhashemi — APPROVED** [review]
Really nice work, the paramrixation of tests is making the tests of these routes so much easier to digest as is the main route logic now that the browse routes are split out!

File: app/templates/main/browse-series.html (hunk 5 above)
Comment (anthonyhashemi) [reply app/templates/main/browse-series.html:149]: Ah yes, I see 👌

File: app/main/routes.py [hunk 8]
```diff
@@ -92,106 +96,238 @@ def accessibility():
...
-            4: {"consignment_reference": consignment.ConsignmentReference},
-        }
-
-        filters = build_browse_consignment_filters(request.args)
-        sorting_orders = build_sorting_orders(request.args)
+        total_records = db.session.query(
+            func.sum(query.subquery().c.records_held)
+        ).scalar()
+        if total_records:
+            num_records_found = total_records
```
Comment (anthonyhashemi) [inline app/main/routes.py:136]: (same as comment on search PR) Isnt the only option here for query to be a positive integer or 0 so we should be fine by just doing `num_record_found = db.session.query(
            func.sum(query.subquery().c.records_held)
        ).scalar()` or could this return a different falsey value such as None?

File: app/tests/test_browse_consignment.py (hunk 7 above)
Comment (pgandhizaizi) [reply app/tests/test_browse_consignment.py:310]: yes, I agree, we can perform sort and filter combinations on lower level

File: app/main/routes.py (hunk 8 above)
Comment (pgandhizaizi) [reply app/main/routes.py:136]: refactored num_record_found as requested

File: app/main/routes.py (hunk 1 above)
Comment (pgandhizaizi) [reply app/main/routes.py:158]: as discussed on slack 'id' keyword shadow 

image

Comment (anthonyhashemi) [reply app/main/routes.py:158]: for others, id is a built in python function https://docs.python.org/3/library/functions.html.

I think due to its low usage in python code it is often as a variable name but making it distinct here avoids possible confusion 👌

File: app/main/routes.py (hunk 8 above)
Comment (pgandhizaizi) [reply app/main/routes.py:136]: refactored num_record_found as it was before

File: app/main/routes.py (hunk 4 above)
Comment (pgandhizaizi) [reply app/main/routes.py:312]: each browse view has been set with default sort order for consistency

Comment (anthonyhashemi) [reply app/main/routes.py:312]: ended up setting the default sort options in each view, so not creating ticket 👍
